// bigbrother-webpack-plugin/lib/BigBrotherWebpackPlugin.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var path_1 = require("path");
var assert = require("assert");
var uuid_1 = require("uuid");
var satori_1 = require("./adapter/satori");
var utils_1 = require("./utils");
var debug = require("debug");
var log = debug('bigbrother-webpack-plugin:BigBrotherWebpackPlugin');
var BigBrotherWebpackPlugin = /** @class */ (function () {
    function BigBrotherWebpackPlugin(options) {
        assert(options, 'options is required');
        this.cwd = options.cwd;
        assert(this.cwd, 'options.cwd is required');
        this.tool = options.tool;
        assert(this.tool, 'options.tool is required');
        this.buildConfig = options.buildConfig || {};
        this.stagesPath = options.stagesPath || path_1.join(this.cwd, 'node_modules', '.build-statistics', 'compilation.json');
        try {
            this.satori = new satori_1.default();
        }
        catch (_) {
            log(_);
        }
    }
    BigBrotherWebpackPlugin.prototype.apply = function (compiler) {
        var _this = this;
        if (compiler.hooks) {
            compiler.hooks.done.tapAsync('monitor', function (stats, callback) {
                // 若有错误，不上传依赖树，需要延迟一秒（防止进程退出导致构建失败信息上报不了）
                if (stats && stats.hasErrors()) {
                    _this.simepleReport(stats);
                    setTimeout(callback, 1000);
                    return;
                }
                process.env.NODE_ENV === 'production'
                    ? _this.report(stats)
                    : _this.simepleReport(stats);
                callback();
            });
            return;
        }
        // webpack 1
        compiler.plugin('done', function (stats) {
            _this.simepleReport(stats);
        });
    };
    BigBrotherWebpackPlugin.prototype.report = function (stats) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var data, fun, _1;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 2, , 3]);
                        return [4 /*yield*/, this.getReportData(stats)];
                    case 1:
                        data = _a.sent();
                        fun = process.env.NODE_ENV === 'development' ? 'postDev' : 'postBuild';
                        this.satori[fun](data);
                        return [3 /*break*/, 3];
                    case 2:
                        _1 = _a.sent();
                        log(_1);
                        return [3 /*break*/, 3];
                    case 3: return [2 /*return*/];
                }
            });
        });
    };
    // 不需要 deps tree 的 report
    BigBrotherWebpackPlugin.prototype.simepleReport = function (stats) {
        var _a = process.versions, node = _a.node, _b = _a.alinode, alinode = _b === void 0 ? '' : _b;
        try {
            var data = {
                buildTimeRef: this.getUniqBuildId(),
                cwd: this.cwd,
                success: !stats.hasErrors(),
                startedAt: stats.startTime,
                endedAt: stats.endTime,
                tool: this.tool,
                pkg: utils_1.getPkg(this.cwd),
                buildConfig: this.buildConfig,
                artifact: utils_1.getArtifactInfo(stats),
                errors: stats.compilation.errors.slice(0, 2).map(function (_a) {
                    var name = _a.name, _b = _a.message, message = _b === void 0 ? '' : _b;
                    return ({
                        name: name,
                        message: message.substr(0, 200),
                    });
                }),
                stages: this.getStages(stats),
                node: {
                    node: node,
                    alinode: alinode || 'N/A',
                },
                git: utils_1.getGitInfo(this.cwd),
                platform: utils_1.getPlatformInfo(),
                envs: process.env,
                hardware: utils_1.getHardware(),
                shameimaruUrl: 'N/A',
            };
            var fun = process.env.NODE_ENV === 'development' ? 'postDev' : 'postBuild'; // 若 NODE_ENV 为空，以 build 处理
            this.satori[fun](data);
        }
        catch (_) {
            log(_);
        }
    };
    BigBrotherWebpackPlugin.prototype.getReportData = function (stats) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var _a, node, _b, alinode, data, depTreeOssUrl, depTree, _2;
            return tslib_1.__generator(this, function (_c) {
                switch (_c.label) {
                    case 0:
                        _a = process.versions, node = _a.node, _b = _a.alinode, alinode = _b === void 0 ? '' : _b;
                        data = {
                            buildTimeRef: this.getUniqBuildId(),
                            cwd: this.cwd,
                            success: !stats.hasErrors(),
                            startedAt: stats.startTime,
                            endedAt: stats.endTime,
                            tool: this.tool,
                            pkg: utils_1.getPkg(this.cwd),
                            buildConfig: this.buildConfig,
                            artifact: utils_1.getArtifactInfo(stats),
                            errors: stats.compilation.errors.slice(0, 2).map(function (_a) {
                                var name = _a.name, _b = _a.message, message = _b === void 0 ? '' : _b;
                                return ({
                                    name: name,
                                    message: message.substr(0, 200),
                                });
                            }),
                            stages: this.getStages(stats),
                            node: {
                                node: node,
                                alinode: alinode || 'N/A',
                            },
                            git: utils_1.getGitInfo(this.cwd),
                            platform: utils_1.getPlatformInfo(),
                            envs: process.env,
                            hardware: utils_1.getHardware(),
                            shameimaruUrl: '',
                        };
                        _c.label = 1;
                    case 1:
                        _c.trys.push([1, 5, , 6]);
                        if (!data.platform.type) return [3 /*break*/, 4];
                        return [4 /*yield*/, utils_1.getDepTree(this.cwd)];
                    case 2:
                        depTree = _c.sent();
                        return [4 /*yield*/, this.satori.uploadShameimaru(depTree)];
                    case 3:
                        depTreeOssUrl = _c.sent();
                        _c.label = 4;
                    case 4: return [3 /*break*/, 6];
                    case 5:
                        _2 = _c.sent();
                        log(_2);
                        depTreeOssUrl = '';
                        return [3 /*break*/, 6];
                    case 6:
                        data.shameimaruUrl = depTreeOssUrl;
                        return [2 /*return*/, data];
                }
            });
        });
    };
    // 取 stages
    BigBrotherWebpackPlugin.prototype.getStages = function (stats) {
        var hash = stats.hash;
        var stages = [];
        var stagesPath = this.stagesPath;
        try {
            var timeline = require(stagesPath);
            stages = timeline.hash === hash ? timeline.stages : [];
        }
        catch (_) {
            log(_);
        }
        return stages;
    };
    BigBrotherWebpackPlugin.prototype.getUniqBuildId = function () {
        return process.env.CLOUD_BUILD_ID || uuid_1.v4();
    };
    return BigBrotherWebpackPlugin;
}());
exports.default = BigBrotherWebpackPlugin;
